### PR TITLE
Fixed bugs in reversion anonymization and add posibility to delete al…

### PR DESCRIPTION
**Pull request by matllubos**
…l versions

### CODE REVIEW

File: gdpr/anonymizers/model_anonymizers.py [hunk 1]
```diff
@@ -55,6 +58,8 @@ def __new__(cls, name, bases, attrs):
 
         if not getattr(new_obj.Meta, 'abstract', False):
             anonymizer_register.register(new_obj.Meta.model, new_obj)
+            new_obj.Meta.anonymize_reversion = getattr(new_obj.Meta, 'anonymize_reversion', False)
+            new_obj.Meta.delete_reversion = getattr(new_obj.Meta, 'delete_reversion', False)
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:62]: why use getattr when we can set default value in metamodel

File: gdpr/anonymizers/model_anonymizers.py [hunk 2]
```diff
@@ -115,10 +120,27 @@ def is_reversible(self, obj) -> bool:
...
+                    for name in field_names
+                    if name in get_reversion_local_field_dict(version)
+                }
+            )
+            for version in versions
+        ]
+        for version, version_dict in versions_update_data:
+            self._perform_version_update(version, version_dict)
+
+    def delete_reversion(self, obj, anonymization:bool):
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py]: added posibility to delete reversion (it is faster and more secure)

File: gdpr/anonymizers/model_anonymizers.py [hunk 3]
```diff
@@ -115,10 +120,27 @@ def is_reversible(self, obj) -> bool:
...
+    def anonymize_reversion(self, obj, field_names, anonymization:bool):
+        versions: List[Version] = get_all_obj_and_parent_versions(obj)
+        versions_update_data = [
+            (
+                version,
+                {
+                    name: self.get_value_from_version(self[name], obj, version, name,
+                                                      anonymization=anonymization)
+                    for name in field_names
+                    if name in get_reversion_local_field_dict(version)
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:133]: there was bug. versions.field_dict raises error if parent object doesn't exists. There must be used only local fields from version

File: gdpr/anonymizers/model_anonymizers.py [hunk 4]
```diff
@@ -169,28 +191,6 @@ def update_field_as_anonymized(self, obj: Model, name: str, legal_reason: Option
                 field=name, is_active=True, content_type=self.content_type, object_id=str(obj.pk)
             ).delete()
 
-    def mark_field_as_anonymized(self, obj: Model, name: str, legal_reason: Optional[LegalReason] = None):
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:172]: I don't know purpose of these methods. These were not tested. And are not used too.

File: gdpr/anonymizers/model_anonymizers.py [hunk 5]
```diff
@@ -199,54 +199,9 @@ def _perform_update(self, obj: Model, updated_data: dict, legal_reason: Optional
         for field_name in updated_data.keys():
             self.update_field_as_anonymized(obj, field_name, legal_reason, anonymization=anonymization)
 
-    def get_parent_models(self, model_or_obj: Union[Model, Type[Model]]) -> List[Type[Model]]:
-        """From model get all it's parent models."""
-        return model_or_obj._meta.get_parent_list()
-
-    def get_all_parent_objects(self, obj: Model) -> List[Model]:
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:206]: moved to functions in utils

File: gdpr/anonymizers/model_anonymizers.py [hunk 6]
```diff
@@ -199,54 +199,9 @@ def _perform_update(self, obj: Model, updated_data: dict, legal_reason: Optional
...
-        return [i for i in parent_objects if i is not None]
-
-    def get_reversion_versions(self, obj: Model):
-        from gdpr.utils import get_reversion_versions
-        versions = [i for i in get_reversion_versions(obj)]  # QuerySet to list
-        parent_obj_versions = [get_reversion_versions(i) for i in self.get_all_parent_objects(obj)]
-        versions += [item for sublist in parent_obj_versions for item in sublist]
-        return versions
-
-    def _perform_anonymization(self, obj: Model, updated_data: dict,
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:230]: again no purpose

File: gdpr/anonymizers/model_anonymizers.py [hunk 7]
```diff
@@ -270,27 +225,6 @@ def _perform_version_update(version, update_data):
         )
         version.save()
 
-    def perform_update_with_version(self, obj: Model, updated_data: Dict[str, Any],
-                                    updated_version_data: List[Tuple[Any, Dict]],
-                                    legal_reason: Optional[LegalReason] = None,
-                                    anonymization: bool = True):
-        with transaction.atomic():
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:277]: the transaction atomic here hase no reason. Whole anonymization from source object must be in atomic block but not here

File: gdpr/anonymizers/model_anonymizers.py [hunk 8]
```diff
@@ -452,28 +386,11 @@ def update_obj(self, obj: Model, legal_reason: Optional[LegalReason] = None,
...
-                    obj, update_dict, versions_update_dict, legal_reason,
-                    anonymization=anonymization
-                )
-            else:
-                self.perform_update(obj, update_dict, legal_reason, anonymization=anonymization)
+            if self.Meta.delete_reversion:
+                self.delete_reversion(obj, anonymization)
+            elif self.Meta.anonymize_reversion:
+                self.anonymize_reversion(obj, raw_local_fields, anonymization)
+            self.perform_update(obj, update_dict, legal_reason, anonymization=anonymization)
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:394]: refactorization

File: gdpr/anonymizers/model_anonymizers.py [hunk 9]
```diff
@@ -522,10 +439,8 @@ def update_obj(self, obj: Model, legal_reason: Optional[LegalReason] = None,
 
             obj.__class__.objects.filter(pk=obj.pk).delete()
 
-            if self.anonymize_reversion(obj):
-                from reversion.models import Version
-                from gdpr.utils import get_reversion_versions
-                get_reversion_versions(obj).delete()
```
Comment (matllubos) [inline gdpr/anonymizers/model_anonymizers.py:528]: parent object versions were not removed

File: gdpr/management/commands/deactivate_expired_reasons.py [hunk 10]
```diff
@@ -11,5 +11,8 @@ class Command(BaseCommand):
 
     def handle(self, *args, **options):
         self.stdout.write('Anonymize expired data of expired legal reasons')
-        for legal_reason in tqdm(LegalReason.objects.filter_active_and_expired()):
+        legal_reason_to_expire_qs = LegalReason.objects.filter_active_and_expired()
+        for legal_reason in tqdm(legal_reason_to_expire_qs.iterator(),
+                                 total=legal_reason_to_expire_qs.count(),
+                                 file=self.stdout):
```
Comment (matllubos) [inline gdpr/management/commands/deactivate_expired_reasons.py:17]: only fixed log and used iterator

File: gdpr/utils.py [hunk 11]
```diff
@@ -75,3 +75,34 @@ def is_reversion_installed():
         return True
     except ImportError:
         return False
+
+
+def get_all_parent_objects(obj: Model) -> List[Model]:
```
Comment (matllubos) [inline gdpr/utils.py:80]: helpers moved from model

File: gdpr/utils.py [hunk 12]
```diff
@@ -75,3 +75,34 @@ def is_reversion_installed():
...
+    for parent_path in parent_paths:
+        parent_obj = obj
+        for path in parent_path:
+            parent_obj = getattr(parent_obj, path, None)
+        parent_objects.append(parent_obj)
+
+    return [i for i in parent_objects if i is not None]
+
+
+def get_all_obj_and_parent_versions_queryset_list(obj: Model) -> List[QuerySet]:
```
Comment (matllubos) [inline gdpr/utils.py:98]: I need all querysets of all versions instance and its parents. Yes I can get it from database and return one big queryset. But there is no API in reversion library to get it. And can be problem if reversion uses NoSQL database (DynamoDB). It is much simplier return querysets in list and delete it one by one.

**Review by rubickcz — APPROVED** [review]

**Pull request merged**